File: cpq-backend/src/test/java/com/cpq/quotation/service/TreeFormulaParityFixtureTest.java
package com.cpq.quotation.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.DynamicTest.dynamicTest;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.Map;

import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

/**
 * task-0803 Task 9-A — 前后端求值引擎（BOM 父子取值 tree_ref / tree_attr）逐位比对。
 *
 * <p><b>为什么需要这份测试</b>：{@code TreeFormulaEvalTest} 只单测了 {@link FormulaCalculator#evaluateExpression}
 * 这一层（人工搭 {@code TreeEvalContext}，绕开了 {@link FormulaCalculator#calculate} 的整页签单元格拓扑排序 +
 * 环检测 + effKey 产出流程）。前端 {@code treeFormula.test.ts} 同理只单测了
 * {@code computeTabFormulasTree} 本身，两边各自独立断言各自的期望值，从未真正对比过"同一份输入，
 * 两套引擎算出的数是否逐位相同"。
 *
 * <p><b>本测试改走两端真正的整页签入口</b>：后端 {@link FormulaCalculator#calculate}（本类），
 * 前端 {@code computeTabFormulasTree}（{@code treeFormulaParityFixture.test.ts}），共享同一份
 * JSON 夹具 {@code dev-docs/task-0803-BOM页签增加父子取值公式/fixtures/tree-formula-parity-cases.json}。
 *
 * <p><b>夹具是唯一真相</b>：本文件与前端测试都从磁盘读取<b>同一个物理文件</b>（不复制），
 * 修改用例只需改一处，两端测试自动同步。找不到该文件视为测试基础设施错误（fail-fast，非静默跳过）。
 *
 * <p>纯 JUnit（不用 {@code @QuarkusTest}）—— BL-0095。
 */
class TreeFormulaParityFixtureTest {

    private static final ObjectMapper M = new ObjectMapper();
    private final FormulaCalculator calc = new FormulaCalculator();

    /**
     * 从 Maven 模块工作目录（通常是 {@code cpq-backend/}）向上找仓库根（同时含 {@code cpq-backend}
     * 与 {@code cpq-frontend} 两个子目录的那一层），再定位共享夹具文件。找不到直接抛异常
     * （测试基础设施问题，不该被吞成"跳过"）。
     */
    private static Path fixturePath() {
        Path dir = Paths.get("").toAbsolutePath();
        for (int i = 0; i < 6 && dir != null; i++) {
            if (Files.isDirectory(dir.resolve("cpq-backend")) && Files.isDirectory(dir.resolve("cpq-frontend"))) {
                return dir.resolve("dev-docs")
                        .resolve("task-0803-BOM页签增加父子取值公式")
                        .resolve("fixtures")
                        .resolve("tree-formula-parity-cases.json");
            }
            dir = dir.getParent();
        }
        throw new IllegalStateException(
            "repo root（同时含 cpq-backend/cpq-frontend 的目录）未找到，起点=" + Paths.get("").toAbsolutePath());
    }

    /** 把夹具的 {@code rows[].values} 包成后端 baseRow 形状：系统列在顶层，业务值在 driverRow 内。 */
    private JsonNode toBaseRows(JsonNode rows) {
        ArrayNode out = M.createArrayNode();
        for (JsonNode r : rows) {
            ObjectNode br = M.createObjectNode();
            JsonNode nodeId = r.get("nodeId");
            JsonNode parentId = r.get("parentId");
            if (nodeId != null && !nodeId.isNull()) br.put("__nodeId", nodeId.asText());
            if (parentId != null && !parentId.isNull()) br.put("__parentId", parentId.asText());
            br.put("__lvl", r.path("lvl").asInt(0));
            br.set("driverRow", r.path("values"));
            br.putObject("basicDataValues");
            out.add(br);
        }
        return out;
    }

    @TestFactory
    Collection<DynamicTest> fixtureTests() throws IOException {
        Path path = fixturePath();
        assertTrue(Files.isRegularFile(path), "共享夹具文件不存在: " + path);
        JsonNode cases = M.readTree(Files.readString(path));
        assertTrue(cases.isArray() && cases.size() > 0, "夹具必须是非空数组");

        Collection<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : cases) {
            String name = c.path("name").asText("(unnamed)");
            JsonNode fields = c.path("fields");
            JsonNode formulas = c.path("formulas");
            JsonNode baseRows = toBaseRows(c.path("rows"));
            JsonNode expected = c.path("expected");

            tests.add(dynamicTest(name, () -> {
                ArrayNode result = calc.calculate(fields, formulas, null, null, baseRows, null,
                    Map.of(), Map.of(), Map.of());

                // rowKey → {field → value}：无 rowKeyFields 时 rowKey = 行下标字符串（"0","1",...），
                // 与夹具 rows 数组下标、以及前端 computeTabFormulasTree 的 Record<number,...> 下标一一对应。
                Map<String, JsonNode> byRowKey = new java.util.HashMap<>();
                for (JsonNode row : result) byRowKey.put(row.path("rowKey").asText(), row.path("values"));

                for (Iterator<Map.Entry<String, JsonNode>> it = expected.fields(); it.hasNext(); ) {
                    Map.Entry<String, JsonNode> rowEntry = it.next();
                    String rowIdx = rowEntry.getKey();
                    JsonNode expectedRow = rowEntry.getValue();
                    JsonNode actualRow = byRowKey.get(rowIdx);
                    assertNotNull(actualRow, "[" + name + "] 行 " + rowIdx + " 在后端结果中缺失（byRowKey keys="
                        + byRowKey.keySet() + "）");
                    for (Iterator<Map.Entry<String, JsonNode>> fit = expectedRow.fields(); fit.hasNext(); ) {
                        Map.Entry<String, JsonNode> fieldEntry = fit.next();
                        String field = fieldEntry.getKey();
                        double expectedVal = fieldEntry.getValue().asDouble();
                        JsonNode actualNode = actualRow.get(field);
                        assertNotNull(actualNode, "[" + name + "] 行 " + rowIdx + " 缺列 [" + field + "]（后端结果="
                            + actualRow + "）");
                        assertEquals(expectedVal, actualNode.asDouble(), 1e-9,
                            "[" + name + "] 行 " + rowIdx + " 列 [" + field + "]");
                    }
                }
            }));
        }
        return tests;
    }
}

File: cpq-frontend/src/pages/quotation/treeFormulaParityFixture.test.ts
import { describe, it, expect } from 'vitest';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  computeTabFormulasTree,
  type ComponentDataItem, type TreeFormulaRowInput,
} from './QuotationStep2';

/**
 * task-0803 Task 9-A — 前后端求值引擎（BOM 父子取值 tree_ref / tree_attr）逐位比对。
 *
 * 为什么需要这份测试（而不是复用 treeFormula.test.ts）：treeFormula.test.ts 只单测前端
 * computeTabFormulasTree 自己的一套硬编码期望值，后端 TreeFormulaEvalTest 同理只单测后端
 * evaluateExpression 层（还绕开了 calculate() 的整页签单元格拓扑/环检测）。两边从未真正对比过
 * "同一份输入，两套独立实现算出的数是否逐位相同"。
 *
 * 本测试与后端 TreeFormulaParityFixtureTest 读取**同一个物理文件**（不复制）：
 *   dev-docs/task-260803-BOM页签增加父子取值公式/fixtures/tree-formula-parity-cases.json
 * 改一处夹具，两端测试同步生效；找不到文件直接抛错（fail-fast，不静默跳过）。
 */

interface FixtureRow {
  nodeId: string;
  parentId: string | null;
  lvl: number;
  values: Record<string, any>;
  /** 可选：建模 BASIC_DATA 字段（键形如 "{$view.col}"）。见后端 harness 同名注释。 */
  basicDataValues?: Record<string, any>;
}

interface FixtureCase {
  name: string;
  _doc?: string;
  fields: { name: string; field_type: string; basic_data_path?: string }[];
  formulas: { name: string; expression: any[] }[];
  rows: FixtureRow[];
  expected: Record<string, Record<string, number>>;
}

/** 从当前测试文件所在目录向上找仓库根（同时含 cpq-backend 与 cpq-frontend 子目录的那一层）。 */
function findRepoRoot(startDir: string): string {
  let dir = startDir;
  for (let i = 0; i < 8; i++) {
    if (fs.existsSync(path.join(dir, 'cpq-backend')) && fs.existsSync(path.join(dir, 'cpq-frontend'))) {
      return dir;
    }
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error(`repo root (containing cpq-backend & cpq-frontend) not found from ${startDir}`);
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const repoRoot = findRepoRoot(__dirname);
const fixturePath = path.join(
  repoRoot, 'dev-docs', 'task-260803-BOM页签增加父子取值公式', 'fixtures', 'tree-formula-parity-cases.json',
);

if (!fs.existsSync(fixturePath)) {
  throw new Error(`共享夹具文件不存在: ${fixturePath}`);
}
const cases: FixtureCase[] = JSON.parse(fs.readFileSync(fixturePath, 'utf-8'));

function buildComp(c: FixtureCase): ComponentDataItem {
  return {
    componentId: 'fixture', componentCode: 'FIXTURE', tabName: 'BOM页签',
    fields: c.fields as any,
    formulas: c.formulas as any,
    rows: [],
    subtotal: 0,
  } as any;
}

function buildRows(c: FixtureCase): TreeFormulaRowInput[] {
  const decimalize = (values?: Record<string, any>) => values && Object.fromEntries(
    Object.entries(values).map(([key, value]) => [key, typeof value === 'number' ? String(value) : value]),
  );
  return c.rows.map(r => ({
    row: decimalize(r.values) ?? {},
    basicDataValues: decimalize(r.basicDataValues),
    nodeId: r.nodeId,
    parentId: r.parentId ?? null,
    lvl: r.lvl,
  }));
}

// fail-fast：夹具必须非空，否则下面 for 循环会静默产出 0 个 it()，测试报告显示"0 passed"而非报错，
// 容易被误读成"全绿"。用顶层断言在 describe 收集阶段就炸，不藏进某个 it 里。
if (cases.length === 0) {
  throw new Error(`共享夹具为空数组: ${fixturePath}`);
}

describe('computeTabFormulasTree — 前后端共享夹具逐位比对（task-0803 Task 9-A）', () => {
  for (const c of cases) {
    it(c.name, () => {
      const comp = buildComp(c);
      const rows = buildRows(c);
      const out = computeTabFormulasTree(comp, rows);

      for (const [rowIdxStr, expectedRow] of Object.entries(c.expected)) {
        const rowIdx = Number(rowIdxStr);
        const actualRow = out[rowIdx];
        expect(actualRow, `[${c.name}] 行 ${rowIdx} 在前端结果中缺失`).toBeDefined();
        for (const [field, expectedVal] of Object.entries(expectedRow)) {
          expect(actualRow[field], `[${c.name}] 行 ${rowIdx} 列 [${field}]`).toBe(String(expectedVal));
        }
      }
    });
  }
});
